File: Packets/packets/StartGame.cs

using System;
using System.Text;


public class StartGame : Packet<PacketType>
{
    public string PlayerId { get; private set; }

    public StartGame(byte[] bytes) : base(bytes) { }

    public StartGame(string playerId, PacketAction action) : base(PacketType.StartGame, action)
    {
        this.PlayerId = playerId;
    }

    protected override void ReadPayload(byte[] bytes, byte offset)
    {
        PlayerId = Encoding.UTF8.GetString(bytes, offset, 8);
    }

    protected override void WritePayload(byte[] bytes, byte offset)
    {
        Array.Copy(Encoding.UTF8.GetBytes(PlayerId), 0, bytes, offset, 8);
    }

    public override string ToString()
    {
        return $"StartGame{{ playerId='{PlayerId}' }}";
    }
}

File: src/main/java/net/nerfatg/proxy/packet/packets/StartGame.java
package net.nerfatg.proxy.packet.packets;

import net.nerfatg.proxy.packet.Packet;
import net.nerfatg.proxy.packet.PacketAction;
import net.nerfatg.proxy.packet.PacketType;

import java.nio.BufferOverflowException;
import java.nio.ByteBuffer;
import java.nio.charset.StandardCharsets;

public class StartGame extends Packet {
    private String playerId;

    public StartGame(ByteBuffer buffer) throws BufferOverflowException {
       super(buffer);
    }

    public StartGame(String playerId, PacketAction action) {
        super(PacketType.StartGame, action);
        this.playerId = playerId;
    }

    @Override
    public void readPayload(ByteBuffer buffer, int offset) throws BufferOverflowException {
        byte[] playerIdBytes = new byte[8];
        buffer.position(offset);
        buffer.get(playerIdBytes);
        this.playerId = new String(playerIdBytes, StandardCharsets.UTF_8);
    }

    @Override
    public void writePayload(ByteBuffer buffer, int offset) throws BufferOverflowException {
        buffer.position(offset);
        
        // Write playerId (8 bytes fixed)
        byte[] originalPlayerIdBytes = playerId.getBytes(StandardCharsets.UTF_8);
        buffer.put(originalPlayerIdBytes, 0, Math.min(originalPlayerIdBytes.length, 8));
        if (originalPlayerIdBytes.length < 8) {
            buffer.put(new byte[8 - originalPlayerIdBytes.length]); // pad with zeros
        }
    }

    public String getPlayerId() {
        return playerId;
    }
}
